pkg/controller/replicaset/replica_set.go

```diff
@@ -542,16 +575,60 @@ func (rsc *ReplicaSetController) syncReplicaSet(key string) error {
 		glog.Errorf("Error converting pod selector to selector: %v", err)
 		return err
 	}
-	podList, err := rsc.podStore.Pods(rs.Namespace).List(selector)
-	if err != nil {
-		glog.Errorf("Error getting pods for ReplicaSet %q: %v", key, err)
-		rsc.queue.Add(key)
-		return err
+
+	// TODO: Do the List and Filter in a single pass, or use an index.
+	var filteredPods []*api.Pod
+	if rsc.garbageCollectorEnabled {
+		// list all pods to include the pods that don't match the rs`s selector
+		// anymore but has the stale controller ref.
+		podList, err := rsc.podStore.Pods(rs.Namespace).List(labels.Everything())
+		if err != nil {
+			glog.Errorf("Error getting pods for rs %q: %v", key, err)
+			rsc.queue.Add(key)
+			return err
+		}
+		cm := controller.NewPodControllerRefManager(rsc.podControl, rs.ObjectMeta, selector, getRSKind())
+		matchesAndControlled, matchesNeedsController, controlledDoesNotMatch := cm.Classify(podList.Items)
+		for _, pod := range matchesNeedsController {
+			err := cm.AdoptPod(pod)
```

Comment:
Yes, both of them builds patch based on the pod and sends the patch to the server.